File: mazes-for-programmers/python/src/mazes/ellers.py
from random import randint, sample

from colored_grid import ColoredGrid
from image_saver import save


class RowState:
    def __init__(self, starting_set=0):
        self.cells_in_set = {}
        self.set_for_cell = {}
        self.next_set = starting_set

    def record(self, cell_set, cell):
        self.set_for_cell[cell.column] = cell_set
        if cell_set not in self.cells_in_set:
            self.cells_in_set[cell_set] = []
        self.cells_in_set[cell_set].append(cell)

    def set_for(self, cell):
        if cell.column not in self.set_for_cell:
            self.record(self.next_set, cell)
            self.next_set += 1
        return self.set_for_cell[cell.column]

    def merge(self, winner, loser):
        for cell in self.cells_in_set[loser]:
            self.set_for_cell[cell.column] = winner
            self.cells_in_set[winner].append(cell)
        del self.cells_in_set[loser]

    def next(self):
        return RowState(self.next_set)

    def each_set(self):
        for cell_set, cells in self.cells_in_set.items():
            yield cell_set, cells


class Ellers:
    @staticmethod
    def on(grid):
        row_state = RowState()
        for row in grid.each_row():
            for cell in row:
                if not cell.west:
                    continue
                cell_set = row_state.set_for(cell)
                prior_set = row_state.set_for(cell.west)
                should_link = cell_set != prior_set and (cell.south is None or randint(0, 1) == 0)
                if should_link:
                    cell.link(cell.west)
                    row_state.merge(prior_set, cell_set)

            if row[0].south:
                next_row = row_state.next()
                for _, cell_list in row_state.each_set():
                    for index, cell in enumerate(sample(cell_list, len(cell_list))):
                        if index == 0 or randint(0, 2) == 0:
                            cell.link(cell.south)
                            next_row.record(row_state.set_for(cell), cell.south)
                row_state = next_row


def ellers_demo():
    grid = ColoredGrid(11, 11)
    Ellers.on(grid)
    save(grid.to_img(), "ellers.png")

    middle = grid[grid.rows // 2, grid.columns // 2]
    grid.set_distances(middle.distances())

    save(grid.to_img(), "ellers_colored.png")


if __name__ == "__main__":
    ellers_demo()

File: mazes-for-programmers/kotlin/src/main/kotlin/io/github/ocirne/mazes/algorithms/Ellers.kt
package io.github.ocirne.mazes.algorithms

import io.github.ocirne.mazes.grids.Cell
import io.github.ocirne.mazes.grids.GridProvider
import io.github.ocirne.mazes.grids.Maze
import io.github.ocirne.mazes.grids.cartesian.CartesianGrid.CartesianCell
import io.github.ocirne.mazes.grids.hex.HexGrid.HexCell
import io.github.ocirne.mazes.grids.polar.PolarGrid.PolarCell
import io.github.ocirne.mazes.grids.triangle.TriangleGrid.TriangleCell
import io.github.ocirne.mazes.grids.upsilon.UpsilonGrid.UpsilonCell
import kotlin.random.Random.Default.nextInt

class Ellers : PassageCarver {

    override fun on(gridProvider: GridProvider, startAt: (Maze) -> Cell): Maze {
        throw NotImplementedError("Please use specialized functions")
    }

    private fun <C : Cell> generalizedEllers(
        grid: GridProvider,
        left: (C) -> C?,
        down: (C) -> List<C>,
        reversedRows: Boolean = false
    ): Maze {
        val maze = grid.forPassageCarver()
        var rowState = RowState()
        for (row in maze.eachRow(reversedRows)) {
            for (cell in row) {
                cell as C
                val currentSet = rowState.setFor(cell)
                val leftCell = left(cell) ?: continue
                val priorSet = rowState.setFor(leftCell)
                val shouldLink = currentSet != priorSet && (down(cell).isEmpty() || nextInt(2) == 0)
                if (shouldLink) {
                    cell.link(leftCell)
                    rowState.merge(priorSet, currentSet)
                }
            }
            if (down(row[0] as C).isNotEmpty()) {
                val nextRow = rowState.next()
                for (cells in rowState.eachSet()) {
                    cells.shuffled().forEachIndexed { index, cell ->
                        cell as C
                        val currentSet = rowState.getSetForCell(cell)!!
                        for (nextCell in down(cell)) {
                            if (index == 0 || nextInt(3) == 0) {
                                if (!nextRow.containsSet(currentSet)) {
                                    cell.link(nextCell)
                                    nextRow.record(rowState.setFor(cell), nextCell)
                                }
                            }
                        }
                    }
                }
                rowState = nextRow
            }
        }
        return maze
    }

    fun onCartesianGrid(grid: GridProvider): Maze {
        return generalizedEllers<CartesianCell>(grid,
            left = { cell -> cell.west },
            down = { cell -> listOfNotNull(cell.south) })
    }

    fun onPolarGridFromCenter(grid: GridProvider): Maze {
        return generalizedEllers<PolarCell>(grid,
            left = { cell -> cell.cw },
            down = { cell -> cell.outward })
    }

    fun onPolarGridFromEdge(grid: GridProvider): Maze {
        return generalizedEllers<PolarCell>(grid,
            left = { cell -> cell.cw },
            down = { cell -> listOfNotNull(cell.inward) },
            reversedRows = true)
    }

    fun onTriangleGrid(grid: GridProvider): Maze {
        TODO("Mazes are not perfect")
        return generalizedEllers<TriangleCell>(grid,
            left = { cell -> cell.west },
            down = { cell -> if (cell.isUpright()) listOfNotNull(cell.south) else listOfNotNull() })
    }

    fun onHexGrid(grid: GridProvider): Maze {
        TODO("Mazes are not perfect")
        return generalizedEllers<HexCell>(grid,
            left = { cell -> cell.northwest },
            down = { cell -> listOfNotNull(cell.southwest, cell.south, cell.southeast) })
    }

    fun onUpsilonGrid(grid: GridProvider): Maze {
        return generalizedEllers<UpsilonCell>(grid,
            left = { cell -> cell.west },
            down = { cell -> listOfNotNull(cell.southeast, cell.south, cell.southwest) })
    }

    internal class RowState(startingSet: Int = 0) {

        private val cellsInSet = mutableMapOf<Int, MutableSet<Cell>>()
        private val setForCell = mutableMapOf<Cell, Int>()
        private var nextSet = startingSet

        fun containsSet(set: Int): Boolean {
            return cellsInSet.contains(set)
        }

        fun getSetForCell(cell: Cell): Int? {
            return setForCell[cell]
        }

        fun record(set: Int, cell: Cell) {
            setForCell[cell] = set
            if (!cellsInSet.contains(set)) {
                cellsInSet[set] = mutableSetOf()
            }
            cellsInSet[set]!!.add(cell)
        }

        fun setFor(cell: Cell): Int {
            if (!setForCell.contains(cell)) {
                record(nextSet, cell)
                nextSet++
            }
            return setForCell[cell]!!
        }

        fun merge(winner: Int, loser: Int) {
            for (cell in cellsInSet[loser]!!) {
                setForCell[cell] = winner
                cellsInSet[winner]!!.add(cell)
            }
            cellsInSet.remove(loser)
        }

        fun next(): RowState {
            return RowState(nextSet)
        }

        fun eachSet(): MutableCollection<MutableSet<Cell>> {
            return cellsInSet.values
        }
    }
}
